README.md

```diff
@@ -1,507 +1,851 @@
-                .and()
-                .formLogin();
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        Optional<User> user = userRepository.findUserByUsername(username);
+
+        // How do we turn our Optional<User> into UserDetails?
     }
 }
 ```
 
-Let's put the 2 versions of the call to `authorizeRequests()` side by side:
+Let's look at the code above.
 
-- http.authorizeRequests().anyRequest().permitAll();
-- http.authorizeRequests().anyRequest().authenticated().and().formLogin();
+We'll have the `UserService` implement the `UserDetailsService`, which means we
+will need to override the `loadUserByUsername()` method. In the method, we'll
+call the `findUserByUsername()` method we created in our repository class. This
+will return an `Optional<User>`.
 
-Let's examine the differences:
+But uh-oh. We aren't supposed to return an `Optional<User>` or even just the
+`User` entity. We need to return a `UserDetails` instance. So how do we do that?
 
-- As you can see, we are applying our new rule to the same types of requests as
-  before by calling `anyRequest()`. Later in this section, we will look at how
-  to apply our rules to a subset of requests.
-- But now instead of asking any request to have the `permitAll()` method apply
-  to it, we are saying that any request actually needs to be `authenticated()`
-  in order to proceed
-- The `and()` after the call to `authenticated()` allows us to chain together
-  all the calls we need in order to configure our Spring Security, rather than
-  having to assign the returns from each method to a variable and then calling
-  additional methods on those variables
-- The next instruction we give to Spring Security is to enable `formLogin()` as
-  the authentication method.
+Let's create a wrapper class that encompasses a `User` instance and implements
+the `UserDetails` interface. Within the `entity` package, create a
+`UserWrapper.java` file that implements the `UserDetails` interface. We'll also
+have to override several methods. Consider the following implementation:
 
-These changes get us back to where we are now challenged with Spring's default
-login page when we try to access our `hello` endpoint:
+```java
+package com.example.springsecuritydemo.entity;
+
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+import java.util.Collection;
+import java.util.List;
+
+public class UserWrapper implements UserDetails {
+
+  private final User user;
+
+  public UserWrapper(User user) {
+    this.user = user;
+  }
+
+  @Override
+  public String getUsername() {
+    return user.getUsername();
+  }
+
+  @Override
+  public String getPassword() {
+    // We'll need to encode the user's password before we return it
+    return new BCryptPasswordEncoder().encode(user.getPassword());
```

Comment:
Could this lesson have more information about what password encoding is and why it's useful?

Reply:
I actually have a completely separate lesson all about encoding, encrypting, and hashing passwords. What if I say:

"We will address password handling and security in a separate lesson later on to better explain this."

^ the first time BCrypt is shown in the lesson?

Reply:
Sure, sounds good.